### ISSUE TITLE
finetune - whisper needs to add EOS token in the labels

### ISSUE THREAD

**Issue by wonkyuml** [huggingface/blog#641]
@sanchit-gandhi Without adding eos token, the model would forget when to stop generating text.

**Comment by sanchit-gandhi** [comment]
The EOS token is added by default by the Whisper tokenizer: https://github.com/huggingface/transformers/pull/19921

You can test this yourself (ensuring you have installed Transformers from main https://huggingface.co/docs/transformers/installation#install-from-source):
```python
from transformers import WhisperTokenizer

tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-small", language="English")

input_str = "the cat"
encoded_ids = tokenizer(input_str).input_ids
decoded_str = tokenizer.decode(encoded_ids, skip_special_tokens=False)
decoded_str_without_special = tokenizer.decode(encoded_ids, skip_special_tokens=True)

print(decoded_str)
print(decoded_str_without_special)
```
**Print Output:**
```
<|startoftranscript|><|en|><|notimestamps|>the cat<|endoftext|>
the cat
```
You can see that `decoded_str` contains the EOS token (`<|endoftext|>`) at the end of the sequence when we decode and retain the 'special' BOS and EOS tokens.

**Comment by RK-BAKU** [comment]
@sanchit-gandhi 

Unfortunatelly, it doesn't work.  Transformers v. 4.24.0
```
tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-base", language="English") # doesn't work with  task="transcribe" either

input_str = "the cat"
encoded_ids = tokenizer(input_str).input_ids
decoded_str = tokenizer.decode(encoded_ids, skip_special_tokens=False)
decoded_str_without_special = tokenizer.decode(encoded_ids, skip_special_tokens=True)

print(decoded_str)
print(decoded_str_without_special)
print(f"Equal:             {decoded_str_without_special == decoded_str}")
```

Returns string without special tokens:

```
the cat
the cat
Equal:             True
```

**Comment by sanchit-gandhi** [comment]
For English we need to make two modifications to the script:
1) Load everything from one of the English only checkpoints ("openai/whisper-base.en"). See the "Introduction" of the blog post for details: https://huggingface.co/blog/fine-tune-whisper#introduction
2) Don't specify `language` or `task` arguments: the English checkpoints only work with English and speech recognition. So omit these arguments when you're using the English only model (the model has learnt this implicitly)

**Comment by RK-BAKU** [comment]
@sanchit-gandhi

Thanks for hints! Now the inference is ok during eval.

Just noticed another "issue"? 
The training is slow - running experiment with batch_size=48 on RTX3090, 24GB RAM.
Got to step 8810 in 18 hours. Is it normal training time for this model? 
Step: 8810/100000 [18:02:37<189:29:41,  7.48s/it
I see

**Comment by sanchit-gandhi** [comment]
Training is roughly twice as long as a CTC model. You shouldn't need much more than 5k steps to get good performance though. 100k steps with a batch_size=48 is quite a lot of training 😅 You can probably reduce this to 10k steps or less

**Comment by RK-BAKU** [comment]
> Training is roughly twice as long as a CTC model. You shouldn't need much more than 5k steps to get good performance though. 100k steps with a batch_size=48 is quite a lot of training 😅 You can probably reduce this to 10k steps or less

I'm training different language on a big dataset, not English.
Could you give some hints for alt language training please?
What training options you'd suggest to finetune, training time, etc?
Thank you for help!

**Comment by RK-BAKU** [comment]
I've already tried the following:

1. not to use dropouts
2. use dropouts:
    attention_dropout=0.05,
    activation_dropout=0.05
3. increase LR x3 as batch_size rose 3 times as well.

But actually I didn't notice any significant WER changes.
Wondering what else could be done in order to improve WER?

**Comment by sanchit-gandhi** [comment]
Hey @RK-BAKU,

To clarify, you're fine-tuning Whisper on a language **other than** English? 

In this case, you need to load the multilingual checkpoint and set the language correctly in the `processor`. Please follow the blog/Colab for instructions on doing this!

> I'm training different language on a big dataset

Two questions! Which language? And how big is 'big'?

If the language is one of the 96 languages Whisper is pre-trained on, you shouldn't require very much fine-tuning whatsoever. The exact amount I can help with if I know the language and size of your fine-tuning dataset.

**Comment by sanchit-gandhi** [comment]
Okay nice! Thanks for answering those questions. Have you tried training Whisper first without any data augmentation (i.e. on the 50h). Here, the model should converge in about 5k train steps.

**Comment by RK-BAKU** [comment]
Thanks! I will try it asap.
